mi: Introduce asynchronous event message handling

chorkin opened the pull request:
Added new functionality to mi.c and mi-mctp.c to handle AEMs.  Included new example mi-mctp-ae.c for usage.

chorkin commented:
@jk-ozlabs, there's something funky going on with the existing mi-mctp tests.  I was hoping you might shed some light here.  Also looking for guidance on how we can add new tests to handle these changes.

chorkin commented:
@igaw, it seems we have build checks that have been added since the last time I upstreamed something.  But they're all failing for me here.  Any guidance you can provide?

chorkin commented on src/nvme/mi.c:
```diff
@@ -1974,3 +2101,497 @@ const char *nvme_mi_status_to_string(int status)
+		goto cleanup_get_info;
+	}
+	//Need to send a AEM Sync to enable things
+#ifndef TEST_CODE
```
@jk-ozlabs , I plan to remove this code, but was going to keep it in if it was somehow useful for testing purposes and we could reuse it somewhere else.  I'll delete or move pending other comments I get.

chorkin commented on src/nvme/mi.c:
```diff
@@ -1974,3 +2101,497 @@ const char *nvme_mi_status_to_string(int status)
+
+/* POLLIN has indicated events, notify libnvme. Will likely result in
+ * one (or more) of the struct nvme_mi_aem_callbacks being invoked.
+ * NOTE: WE NEED TO LOOK AT AEM GENERATION NUMBER HERE.  See AEMTI
```
Stale comment

chorkin commented on src/libnvme-mi.map:
```diff
@@ -1,11 +1,17 @@
 LIBNVME_MI_1_11 {
 	global:
 		nvme_mi_control;
+		nvme_mi_get_async_message;
```
This is a mistake and should be removed

chorkin commented on src/nvme/mi-mctp.c:
```diff
@@ -220,6 +224,179 @@ static bool nvme_mi_mctp_resp_is_mpr(void *buf, size_t len,
+	pollfds[0].events = POLLIN;
+	timeout = 1;//1?
+retry:
+	rc = ops.poll(pollfds, 1, timeout);//This is probably unecessary
```
Stale comment.  Same for //1? above

chorkin commented on src/nvme/mi.c:
```diff
@@ -429,11 +486,6 @@ int nvme_mi_submit(nvme_mi_ep_t ep, struct nvme_mi_req *req,
 		return -1;
 	}
 
-	if (req->data_len & 0x3) {
```
@jk-ozlabs , correct me if I'm wrong, but this check should only be applicable to admin commands and is actually causing problems with AEM SYNC (Configuration Set: AEM) for some data lengths.  Didn't want it to slip under the radar.

jk-ozlabs replied:
Up to this point, all MI messages have been a multiple of 4 bytes. If that's changing, we can reconsider this check.

chorkin replied:
Yes.  For Configuration Set:AE, the message length varies and the amount it can change size is in multiples of 3

jk-ozlabs commented:
Neat, I'll take a look!

jk-ozlabs commented on src/nvme/mi-mctp.c:
```diff
@@ -459,12 +638,15 @@ static const struct nvme_mi_transport nvme_mi_transport_mctp = {
 	.submit = nvme_mi_mctp_submit,
 	.close = nvme_mi_mctp_close,
 	.desc_ep = nvme_mi_mctp_desc_ep,
+	.async_read = nvme_mi_mctp_read,
+	.async_poll_fd = nvme_mi_mctp_async_poll_fd,
```
It's only a minor naming convention, but we probably want to be careful between the terms `async` (as an IO-handling pattern) and "Async Event Messaging" (as the NVMe-MI facility). I would suggest using `aem` consistently here.

jk-ozlabs commented on examples/mi-mctp-ae.c:
```diff
@@ -0,0 +1,186 @@
+		std_in_timeout.tv_usec = 0;
+
+		// Check if there's input on stdin
+		retval = select(STDIN_FILENO + 1, &read_fds, NULL, NULL, &std_in_timeout);
```
Just include `STDIN_FILENO` on the set that you're polling on

jk-ozlabs commented:
I think the overall concept here is sound, but we are missing a few implementation details for an overall feasibility check too. There are some style things, but we can address those once the general framework is settled. I have added some comments in a review.

I've been thinking about how we should be handling the multiple-endpoint cases, as we can only have one socket listening on incoming NVMe-MI message types. A couple of options there:

 * a shared socket for all AEM events on all endpoints (eg, on the `nvme_mi_root_t`), where a call to fetch AEM events on any enpoint will receive all events, and queue those that are not for the current. A subsequent call for any other endpoints would first check that queue before receiving more directly from the socket

or:

 * we add socket syscall semantics for AF_MCTP, to specify the remote MCTP EID on the listening socket, and stick with one per `nvme_mi_ep_t`. this would require kernel changes, but I think that's probably a common use-case that would be good to support.

I figure the latter is probably best, so will take a look at what's required for that.

jk-ozlabs commented:
(BTW, I would suggest converting to a draft while still working out details)

igaw commented:
> @igaw, it seems we have build checks that have been added since the last time I upstreamed something. But they're all failing for me here. Any guidance you can provide?

The build container cleanup task eagerly removed containers.

chorkin commented:
> Collaborator



> I think the overall concept here is sound, but we are missing a few implementation details for an overall feasibility check too. There are some style things, but we can address those once the general framework is settled. I have added some comments in a review.
> 
> I've been thinking about how we should be handling the multiple-endpoint cases, as we can only have one socket listening on incoming NVMe-MI message types. A couple of options there:
> 
> * a shared socket for all AEM events on all endpoints (eg, on the `nvme_mi_root_t`), where a call to fetch AEM events on any enpoint will receive all events, and queue those that are not for the current. A subsequent call for any other endpoints would first check that queue before receiving more directly from the socket
> 
> or:
> 
> * we add socket syscall semantics for AF_MCTP, to specify the remote MCTP EID on the listening socket, and stick with one per `nvme_mi_ep_t`. this would require kernel changes, but I think that's probably a common use-case that would be good to support.
> 
> I figure the latter is probably best, so will take a look at what's required for that.

@jk-ozlabs , I wasn't aware of this limitation.  I think the kernel changes are probably the best solution here, but the case we are talking about is probably not going to be a primary use case.  I was wondering if we can come up with a good strategy here to solve this multi-endpoint AEM issue without holding up the initial PR.  Some thought's I had were:

1. Provide this as a known limitation for the current PR and work on kernel changes in background.  Add a follow up to remove this limitation.
2. Implement the shared socket for a short term solution in this PR so there will be no limitation.  Follow up with kernel changes later to simplify the code and memory usage of the application.  I'll see if I can prototype this if I have time.

chorkin commented on src/nvme/mi-mctp.c:
```diff
@@ -220,6 +224,179 @@ static bool nvme_mi_mctp_resp_is_mpr(void *buf, size_t len,
+	pollfds[0].events = POLLIN;
+	timeout = 1;//1?
+retry:
+	rc = ops.poll(pollfds, 1, timeout);//This is probably unecessary
```
mi.c at ~L428 is checking for EWOULDBLOCK because that's how the code is currently filtering endpoints (see ~L349 of mi-mctp.c).  What do you think about doing the read and returning that code up in this case.  It'll cause the higher layers to just act like there was no data.

chorkin commented on src/nvme/mi.h:
```diff
@@ -3086,4 +3326,140 @@ int nvme_mi_admin_format_nvm(nvme_mi_ctrl_t ctrl,
+ *
+ * Return: -1 if error, 0 otherwise
+ */
+int nvme_mi_get_pollfd(nvme_mi_ep_t ep, struct pollfd *fs);
```
Nope

chorkin commented:
@igaw , I've addressed most of the checkpatch issues, but I can't resolve the packed ones.  When I add __packed the structures aren't actually getting packed correctly.  I also see plenty of usage in the codebase in the style I'm using.

jk-ozlabs commented on src/libnvme-mi.map:
```diff
@@ -1,6 +1,11 @@
+		nvme_mi_aem_process;
+		nvme_mi_disable_aem;
+		nvme_mi_aem_get_next_event;
 };
```
Having all of these together highlights a bit of naming inconsistency: some are verb_scope, others are scope_verb. We generally have scope_verb (but where the verb may sometimes be a nvme command name, so it can get a bit mixed up!).

Given a scope of `nvme_mi_aem`, I would suggest:

```
	nvme_mi_aem_get_fd;
	nvme_mi_aem_enable;
	nvme_mi_aem_process;
	nvme_mi_aem_disable;
	nvme_mi_aem_get_next_event;
```

jk-ozlabs commented on src/nvme/mi-mctp.c:
```diff
@@ -84,6 +85,9 @@ struct nvme_mi_transport_mctp {
 	int	sd;
 	void	*resp_buf;
 	size_t	resp_buf_size;
+	int sd_aem;
+	void	*resp_buf_aem;
+	size_t	resp_buf_aem_size;
```
super minor: alignment here

jk-ozlabs commented on src/nvme/mi-mctp.c:
```diff
@@ -487,13 +665,39 @@ nvme_mi_ep_t nvme_mi_open_mctp(nvme_root_t root, unsigned int netid, __u8 eid)
+		goto err_free_aem_rspbuf;
+	}
+
+	mctp->sd_aem = ops.socket(AF_MCTP, SOCK_DGRAM, 0);
```
any thoughts on lazily creating this socket? (ie, only when the caller enables AEM?)

chorkin replied:
What is the benefit?  It means the code paths are less symmetrical, right?

jk-ozlabs commented on src/nvme/mi-mctp.c:
```diff
@@ -487,13 +665,39 @@ nvme_mi_ep_t nvme_mi_open_mctp(nvme_root_t root, unsigned int netid, __u8 eid)
+	addr.smctp_type = MCTP_TYPE_NVME | MCTP_TYPE_MIC;
+	addr.smctp_tag = MCTP_TAG_OWNER;
+
+	if (bind(mctp->sd_aem, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
```
As mentioned in other comments, this will fail on the second invocation `nvme_mi_open_mctp`; you can only have one bind to the same local address. If we open the socket lazily, that may hide this error a little too. Will discuss separately though.

jk-ozlabs commented on src/nvme/mi.c:
```diff
@@ -414,6 +420,59 @@ static int nvme_mi_verify_resp_mic(struct nvme_mi_resp *resp)
+	int rc = ep->transport->aem_read(ep, resp);
+
+	if (nvme_mi_ep_has_quirk(ep, NVME_QUIRK_MIN_INTER_COMMAND_TIME))
+		nvme_mi_record_resp_time(ep);
```
Good catch. Would be good to validate if this quirk still affects the AEM case, but I suspect those parts do not support AEM anyway.

chorkin replied:
Hahaha.  This is copy paste, but thanks.

jk-ozlabs commented:
> * Provide this as a known limitation for the current PR and work on kernel changes in background.  Add a follow up to remove this limitation.

The issue with this is that the limitation is somewhat hidden, only occurring when you try to open (and `bind()`) a second AEM socket.

> * Implement the shared socket for a short term solution in this PR so there will be no limitation.  Follow up with kernel changes later to simplify the code and memory usage of the application.  I'll see if I can prototype this if I have time.

This would be the safer (and more backwards-compatible) approach, but does mean more work.

For a compromise, depending on your requirements, you may be able to run with a modified kernel while doing development, which would support the multiple binds(). We conditionally would enable the AEM support on a recent-enough kernel, and then don't need to worry about demultiplexing events in libnvme. Would that work?

igaw commented:
> @igaw , I've addressed most of the checkpatch issues, but I can't resolve the packed ones. When I add __packed the structures aren't actually getting packed correctly. I also see plenty of usage in the codebase in the style I'm using.

`__packed` is supposed to work correctly. Last time I looked at an alignment problem, it was doing the right thing.
What do you with `getting packed correctly`? Does it not pack at all? Could you post the pahole output of it?

> I also see kdoc failure. What's the best way for me to dig into what's going on here? Is there a way to build the documentation locally?

The `test` target is running the `kdoc` test case, so `meson test -C .build` should do the trick.

igaw commented:
> > @igaw , I've addressed most of the checkpatch issues, but I can't resolve the packed ones. When I add __packed the structures aren't actually getting packed correctly. I also see plenty of usage in the codebase in the style I'm using.
> 
> `__packed` is supposed to work correctly. Last time I looked at an alignment problem, it was doing the right thing. What do you mean with `getting packed correctly`? Does it not pack at all? Could you post the pahole output of it?

Forget it, the reason why `__packed` doesn't work, is that libnvme doesn't define. I mixed this up with nvme-cli. We should cleanup the mess.

chorkin commented:
> > * Provide this as a known limitation for the current PR and work on kernel changes in background.  Add a follow up to remove this limitation.
> 
> The issue with this is that the limitation is somewhat hidden, only occurring when you try to open (and `bind()`) a second AEM socket.
> 
> > * Implement the shared socket for a short term solution in this PR so there will be no limitation.  Follow up with kernel changes later to simplify the code and memory usage of the application.  I'll see if I can prototype this if I have time.
> 
> This would be the safer (and more backwards-compatible) approach, but does mean more work.
> 
> For a compromise, depending on your requirements, you may be able to run with a modified kernel while doing development, which would support the multiple binds(). We conditionally would enable the AEM support on a recent-enough kernel, and then don't need to worry about demultiplexing events in libnvme. Would that work?

I would say my preference, in order of most preferred to least would be:

1) Implement Lazy Socket initialization and 

> > * Provide this as a known limitation for the current PR and work on kernel changes in background.  Add a follow up to remove this limitation.
> 
> The issue with this is that the limitation is somewhat hidden, only occurring when you try to open (and `bind()`) a second AEM socket.
> 
> > * Implement the shared socket for a short term solution in this PR so there will be no limitation.  Follow up with kernel changes later to simplify the code and memory usage of the application.  I'll see if I can prototype this if I have time.
> 
> This would be the safer (and more backwards-compatible) approach, but does mean more work.
> 
> For a compromise, depending on your requirements, you may be able to run with a modified kernel while doing development, which would support the multiple binds(). We conditionally would enable the AEM support on a recent-enough kernel, and then don't need to worry about demultiplexing events in libnvme. Would that work?

My worry is we won't be able to get the Kernel updates on time.  Right now we're running internally with what is here, so there's no rush to get this right other than that we're introducing a Kernel change dependency.  I also can't control when our Kernel get's this update so that's a bit worrisome.  So I would say my preference is to either:

1) Release this change initially with a lazy socket initialization.  That means the limitation is that there can't be two endpoints that enable AEM, but otherwise you can open multiple endpoints for other purposes (yes?).  Then later we can follow up with kernel updates.

2) Bite the bullet and add the socket at the root and queue up responses.  This removes the kernel dependency entirely but we can always come back to this in the future if that Kernel limitation is no loger there.

chorkin commented:
> > > @igaw , I've addressed most of the checkpatch issues, but I can't resolve the packed ones. When I add __packed the structures aren't actually getting packed correctly. I also see plenty of usage in the codebase in the style I'm using.
> > 
> > 
> > `__packed` is supposed to work correctly. Last time I looked at an alignment problem, it was doing the right thing. What do you mean with `getting packed correctly`? Does it not pack at all? Could you post the pahole output of it?
> 
> Forget it, the reason why `__packed` doesn't work, is that libnvme doesn't define it. I mixed this up with nvme-cli. We should cleanup the mess.

Do you have a suggestion to clean this up, or can this be done as a separate PR?

jk-ozlabs commented:
> 2\. Bite the bullet and add the socket at the root and queue up responses.  This removes the kernel dependency entirely but we can always come back to this in the future if that Kernel limitation is no loger there.

The issue here is that we can not remove this this when we add the kernel support for per-EID binds, otherwise you risk breaking existing use-cases. We will be stuck with having the queueing implementation for a long time (or at least a fallback to queueing).

Not sure I fully understood your timing requirements there - are you not okay with carrying an AF_MCTP patch while it makes its way upstream?

chorkin commented:
> > 2. Bite the bullet and add the socket at the root and queue up responses.  This removes the kernel dependency entirely but we can always come back to this in the future if that Kernel limitation is no longer there.
> 
> The issue here is that we can not remove this this when we add the kernel support for per-EID binds, otherwise you risk breaking existing use-cases. We will be stuck with having the queueing implementation for a long time (or at least a fallback to queueing).
> 
> Not sure I fully understood your timing requirements there - are you not okay with carrying an AF_MCTP patch while it makes its way upstream?

If we abstract how this works under the API, how will it break use cases?  Are you making the case that if someone uses this before this feature is in the upstreamed kernel, and then we change to rely on the kernel their system breaks?

I'm not sure I can carry an AF_MCTP patch at all.  We're working with emulators provided by 3rd parties.  I don't have the means to change the kernel.

jk-ozlabs commented:
> Are you making the case that if someone uses this before this feature is in the upstreamed kernel, and then we change to rely on the kernel their system breaks?

That's correct; if we ever remove the queuing code (that allows this to work on kernels that do not have per-EID binds), this will break users that are still on the old kernel.

> I'm not sure I can carry an AF_MCTP patch at all. We're working with emulators provided by 3rd parties. I don't have the means to change the kernel.

OK, so you'll need the queueing code regardless then.

jk-ozlabs commented:
Or: You could also just keep the queuing part in your own tree, and the upstream version would rely on the per-EID binds, if that option makes future maintenance simpler.

igaw commented:
> > > 
> > 
> > 
> > Forget it, the reason why `__packed` doesn't work, is that libnvme doesn't define it. I mixed this up with nvme-cli. We should cleanup the mess.
> 
> Do you have a suggestion to clean this up, or can this be done as a separate PR?

Let's addresss this separately. Not sure if we actually should as it will make the header situation messy. And as library it should try to stay as clean and quirk free as possible. It's possible to disable some warnings in checkpatch, maybe we can just tell to ignore this one.

Anyway just use the `__attribute__((packed))` form for now.

jk-ozlabs commented:
> Anyway just use the `__attribute__((packed))` form for now.

... but with a caveat that you can't rely on how bitfields are laid out in a struct; you may want to define accessors for those.

chorkin commented:
> Or: You could also just keep the queuing part in your own tree, and the upstream version would rely on the per-EID binds, if that option makes future maintenance simpler.

Here's a thought.  What if we leverage the __attribute__((weak)) attribute.  We can implement the weak function to leverage the kernel changes, but we (myself or another user) could implement their own 'strong' function to deal with the queueing?  Might be a bit hard to make it work with the private headers, but could be worth a shot.  With this approach the upstream code matches my tree, but I can implement something now.

jk-ozlabs commented:
> Here's a thought. What if we leverage the **attribute**((weak)) attribute. We can implement the weak function to leverage the kernel changes, but we (myself or another user) could implement their own 'strong' function to deal with the queueing?

I would prefer not - with this, you're leaking some highly-internal details of libnvme out to the library user, and so those details are now subject to backwards compat constraints.

jk-ozlabs commented:
How about:

 1. you implement the per-endpoint socket, as you had planned initially. This won't really be suitable for direct upstreaming with the one-device limitation, but sounds like it would be suitable for your use-case.
 2. I add kernel support for the per-EID bind()s, and add that support to your libnvme changes in (1)
 3. Once we have (2), that is what we propose for upstream libnvme
 4. If you need a multi-endpoint implementation with an old kernel, you can add the queueing in your own patchset

Would that suit?

This way:
 -  we don't need to carry the queueing implementation forever,
 - upstream libnvme stays fairly clean (and we document that the AEM support is dependent on the kernel changes)
 - you can carry on with the simpler single-endpoint implementation (assuming that suits your aims there!)

chorkin commented:
> How about:
> 
> 1. you implement the per-endpoint socket, as you had planned initially. This won't really be suitable for direct upstreaming with the one-device limitation, but sounds like it would be suitable for your use-case.
> 2. I add kernel support for the per-EID bind()s, and add that support to your libnvme changes in (1)
> 3. Once we have (2), that is what we propose for upstream libnvme
> 4. If you need a multi-endpoint implementation with an old kernel, you can add the queueing in your own patchset
> 
> Would that suit?
> 
> This way:
> 
> * we don't need to carry the queueing implementation forever,
> * upstream libnvme stays fairly clean (and we document that the AEM support is dependent on the kernel changes)
> * you can carry on with the simpler single-endpoint implementation (assuming that suits your aims there!)

Unfortunately this isn't going to work.  While we are using only one device right now, we will likely have multiple devices longer term.  I went and checked on how often we can update the kernel and it just doesn't seem like we'll be able to bring in kernel changes in the right timeframe.  It seems this only leaves us with the option to build in the queue now and maintain it in case of early adopters.

chorkin commented on src/nvme/mi-mctp.c:
```diff
@@ -487,13 +693,43 @@ nvme_mi_ep_t nvme_mi_open_mctp(nvme_root_t root, unsigned int netid, __u8 eid)
+	//Skip this for now for test coverage
+	//because it messes with the test_peer SD
+	if (!(mctp->net == 0 && mctp->eid == 0)) {
+		mctp->sd_aem = ops.socket(AF_MCTP, SOCK_DGRAM, 0);
```
@jk-ozlabs , this was the code causing the existing tests to fail.  Because the ops overload for tests is assigning a socket to the test_peer static structure, the first SD does in fact match, but then when we assign a second SD here it means the test_peer no longer matches the mctp->sd since test_peer gets updated again. 

I'm not sure what the best way to solve this will be.

jk-ozlabs replied:
Sounds like we'd nee the test code to be capable of handling multiple socket instances (and the new socket calls required for the AEM socket too). Something like a simple fd table might suit.

chorkin replied:
Okay. I've resolved this with an array, plus a new OPS function for bind.  I'm wondering what kind of tests you envision?  High level tests on the exposed APIs, Low Level Unit tests, or a combination?  I'm a little hesitant to work on this bit until we have the non-test code in a more final state.

igaw commented:
> > How about:
> > 
> > 1. you implement the per-endpoint socket, as you had planned initially. This won't really be suitable for direct upstreaming with the one-device limitation, but sounds like it would be suitable for your use-case.
> > 2. I add kernel support for the per-EID bind()s, and add that support to your libnvme changes in (1)
> > 3. Once we have (2), that is what we propose for upstream libnvme
> > 4. If you need a multi-endpoint implementation with an old kernel, you can add the queueing in your own patchset
> > 
> > Would that suit?
> > This way:
> > 
> > * we don't need to carry the queueing implementation forever,
> > * upstream libnvme stays fairly clean (and we document that the AEM support is dependent on the kernel changes)
> > * you can carry on with the simpler single-endpoint implementation (assuming that suits your aims there!)
> 
> Unfortunately this isn't going to work. While we are using only one device right now, we will likely have multiple devices longer term. I went and checked on how often we can update the kernel and it just doesn't seem like we'll be able to bring in kernel changes in the right timeframe. It seems this only leaves us with the option to build in the queue now and maintain it in case of early adopters.

With my upstream hat on, I say the project should focus on getting it's own future correct. Any downstream user needs to backport/maintain the features/patches it needs themself. Pushing downstream limitation to upstream project is not a good strategy in the long run. 

Thus, I support @jk-ozlabs plan how to add support for the multi-endpoint feature. Get the kernel support added, update the library.

chorkin commented:
> > > How about:
> > > 
> > > 1. you implement the per-endpoint socket, as you had planned initially. This won't really be suitable for direct upstreaming with the one-device limitation, but sounds like it would be suitable for your use-case.
> > > 2. I add kernel support for the per-EID bind()s, and add that support to your libnvme changes in (1)
> > > 3. Once we have (2), that is what we propose for upstream libnvme
> > > 4. If you need a multi-endpoint implementation with an old kernel, you can add the queueing in your own patchset
> > > 
> > > Would that suit?
> > > This way:
> > > 
> > > * we don't need to carry the queueing implementation forever,
> > > * upstream libnvme stays fairly clean (and we document that the AEM support is dependent on the kernel changes)
> > > * you can carry on with the simpler single-endpoint implementation (assuming that suits your aims there!)
> > 
> > 
> > Unfortunately this isn't going to work. While we are using only one device right now, we will likely have multiple devices longer term. I went and checked on how often we can update the kernel and it just doesn't seem like we'll be able to bring in kernel changes in the right timeframe. It seems this only leaves us with the option to build in the queue now and maintain it in case of early adopters.
> 
> With my upstream hat on, I say the project should focus on getting it's own future correct. Any downstream user needs to backport/maintain the features/patches it needs themself. Pushing downstream limitation to upstream project is not a good strategy in the long run.
> 
> Thus, I support @jk-ozlabs plan how to add support for the multi-endpoint feature. Get the kernel support added, update the library.

OK.  In the meantime, is there anything else I can help with in this PR?  It seems like the action is mostly on @jk-ozlabs  at this point.  I'm going to wait to make any mi-mctp test changes because I think we should wait to see what Jeremy does here first. 

Also, regarding mi test changes, I can work on this if it make sense, but am looking for some suggestions from Jeremy on how detailed we want to make it.  As is the APIs will make multiple NVMe-MI transfers to do things and it may be a bit hacky to be able to test this end to end without a target.  We can definitely do it, but it will probably be a good deal of work.  So I want to make sure we agree on the approach first.  A basic example is the enable.  That should be sending a GET CONFIG, then a SET CONFIG (as currently implemented).

jk-ozlabs commented:
> It seems like the action is mostly on @jk-ozlabs at this point

Just on the multiple-binds functionality, or are there other parts that you'd like me to look at?

(or is this mainly on the review aspect?)

chorkin commented:
> > It seems like the action is mostly on @jk-ozlabs at this point
> 
> Just on the multiple-binds functionality, or are there other parts that you'd like me to look at?
> 
> (or is this mainly on the review aspect?)

HI @jk-ozlabs.  Will wait on @igaw  final response regarding queuing [here](https://github.com/linux-nvme/libnvme/pull/967#issuecomment-2715109573) but I would say if we are not going to bring these changes in with queuing, I'm basically at the end of what I can do on this change request without further feedback or the changes to the kernel.  Even when the changes to the kernel are there, I'm not sure I can test those in my environment.  So I guess if you have any final feedback, I can make updates, and if not, I'm looking for a timeline of when you think you can get these upstreamed to the kernel and then make appropriate updates to this PR so that it can get closed out.

igaw commented:
My previous response was on how things should general work. Obviously, not everything is 1 or 0, there is leeway. If your approach is not hindering the future libnvme development or dropping features later on, it could be considered. But
I am not maintaining the `mi` bits, so it's Jeremy's call.

jk-ozlabs commented:
> I'm basically at the end of what I can do on this change request without further feedback or the changes to the kernel

You're still able to implement all of the bits here, which will still work in the one-endpoint scenario for your development work.

Once we have the kernel changes in place, I will contribute a patch to you that adds a small change to use the new kernel feature, to allow multi-endpoint functionality. At that point, we should be good for a merge.

chorkin commented:
> > I'm basically at the end of what I can do on this change request without further feedback or the changes to the kernel
> 
> You're still able to implement all of the bits here, which will still work in the one-endpoint scenario for your development work.
> 
> Once we have the kernel changes in place, I will contribute a patch to you that adds a small change to use the new kernel feature, to allow multi-endpoint functionality. At that point, we should be good for a merge.

Okay.  So as far as I know, I only have tests to add.  Here's my plan, please let me know if I'm overdoing or underdoing it:

- API Level happy path tests for enable/disable/get_enabled which will listen to tx from LibNvme and validate as well as send rx to test upper layer handling.  Plan to build this into test/mi-mctp.c due to the infrastructure of tx/rx, but happy to make a new test module if you feel that's better.
- API Level validation failure tests where caller has issues (ex: not enough buffer)
- API Level validation failure tests where endpoint has provided buggy responses

Does that sound good?  Anything else you are expecting from me on this @jk-ozlabs ?

jk-ozlabs commented:
> So as far as I know, I only have tests to add.

okay, neat! I'm fine to do a more thorough review without those in place, so feel free to develop concurrently, but I don't want to block things on that at this stage. Would be good to have those sorted for merge if possible.

I'm happy to help out with some test strategies if you like; we may need some extra test infrastructure for the new socket calls.

@igaw : the CI failures here are just `__attribute__((packed))` vs. `__packed`. Are you okay to merge like that, or should we sort out the checkpatch configuration beforehand?

I'm just on my way back home from a conf trip, will check out the code locally and get a full check done early next week.

igaw commented:
I am fine with it, it's your call. 

`__attribute__((packed))`: I've just updated the checkpatch configuration, it should not warn about this anymore.

jk-ozlabs commented:
OK, perfect!

@chorkin, if you can rebase onto the latest, those checkpatch failures should disappear :)

jk-ozlabs commented on src/nvme/mi-mctp.c:
```diff
@@ -97,6 +102,7 @@ static struct __mi_mctp_socket_ops ops = {
 	recvmsg,
 	poll,
 	ioctl_tag,
+	bind,
```
These socket ops are only here for the test framework; with the two socket types now (the message fd, and the aem fd), we might want to shift the abstraction here a little.

For example, we could do something like:

```diff
 static struct __mi_mctp_socket_ops ops = {
-	socket,
+ 	msg_socket,
+       aem_socket,
 	sendmsg,
 	recvmsg,
 	poll,
 	ioctl_tag,
};
```

where `aem_socket` and `msg_socket` provide the ops for creating and setting-up the appropriate socket fd. For the AEM case, that would handle the bind() as well.

This may mean it's easier to do the socket handling for the aem tests, as we know which type of mock socket to create.

jk-ozlabs commented on src/nvme/mi-mctp.c:
```diff
@@ -220,6 +226,200 @@ static bool nvme_mi_mctp_resp_is_mpr(void *buf, size_t len,
+	if (ep->transport != &nvme_mi_transport_mctp) {
+		errno = EINVAL;
+		return -1;
+	}
```
now we know it's a mctp transport, just assign to `mctp` here, and you don't need to refer to `->transport_data` multiple times below.

jk-ozlabs commented on src/nvme/mi-mctp.c:
```diff
@@ -220,6 +226,200 @@ static bool nvme_mi_mctp_resp_is_mpr(void *buf, size_t len,
+
+	struct nvme_mi_transport_mctp *mctp = ep->transport_data;
+
+	// Set the socket to non-blocking mode
```
Should it not be in non-blocking mode already? Otherwise your aem_read() may block, which is not what you want.

jk-ozlabs commented on src/nvme/mi-mctp.c:
```diff
@@ -220,6 +226,200 @@ static bool nvme_mi_mctp_resp_is_mpr(void *buf, size_t len,
+	pollfds[0].events = POLLIN;
+	timeout = 1;
+retry:
+	rc = ops.poll(pollfds, 1, timeout);
```
No need to poll on the AEM socket, since we expect the caller to have already done so - just handle the `EAGAIN` on the recvmsg.

jk-ozlabs commented on src/nvme/mi.c:
```diff
@@ -21,6 +21,12 @@
+#define NUM_ENABLES    (256u)
+
+static int nvme_mi_get_async_message(nvme_mi_ep_t ep,
+							struct nvme_mi_aem_msg *aem_msg,
+							size_t *aem_msg_len);
```
Minor, but there's a bit of weird formatting going on here (and `get_async_message`). Do you have tabs set to 8 columns?

chorkin replied:
No?

jk-ozlabs replied:
ok, you should set to 8 then. The continuations should be aligned under the first argument:

```c
static int nvme_mi_get_async_message(nvme_mi_ep_t ep,
				     struct nvme_mi_aem_msg *aem_msg,
				     size_t *aem_msg_len);
```

or wrap to 100 at a single indent:

```c
static int nvme_mi_get_async_message(nvme_mi_ep_t ep,
	struct nvme_mi_aem_msg *aem_msg, size_t *aem_msg_len);
```

(@igaw do you have a preferred style here?)

igaw replied:
Not really. We have both styles in the code base and it depends on the case which one works better.
I'd say use the one which looks sane.

Another option would be to pick one and use it everywhere and fixup the code base. In this case, I would use
the second one but with two tabs, so that the variable definition in the block below don't align with the args. 
I usually find it easier to read. But this is a personal style thing and this is not my code base, just looking
after it. Thus we as group should figure out what makes the most sense in the long run.

jk-ozlabs commented on src/nvme/mi.h:
```diff
@@ -89,6 +89,7 @@
 
 #include "types.h"
 #include "tree.h"
+#include <poll.h>
```
I don't think we need this? (and if we do, it might indicate a layering violation)

jk-ozlabs commented on src/nvme/mi.h:
```diff
@@ -307,6 +310,173 @@ enum nvme_mi_config_smbus_freq {
+ */
+struct ae_supported_item_t {
+	__u8 aesl;//Length of this item.  Set to 3
+	struct {
+		__u16 aeis     : 8; //Identifier of supported ae
+		__u16 reserved : 7;
+		__u16 aese     : 1; //AE Support Enabled
+	} __attribute__((packed)) aesi;
```
Some notes on this one, but they apply to all the new structs that represent on-the-wire data:

The layout of this struct is compiler-implementation-dependent, which may or may not match the format on the wire. I would recommend avoiding bitfields in packed structs (here and elsewhere).

Please keep the `nvme_mi` prefix too; given this is a publicly-accessible header, we want all definitions to be properly "namespaced". We use `ae` here, should that be consistent with the `aem` names used elsewhere?

For anything that is wider than 8 bits, please use the endian annotated types. (`__le32` / `__le16`, I suspect).

chorkin replied:
Understood on the prefixes/namespaces.  Also for le.  Can you let me know how you suggest I handle bitfield?

jk-ozlabs commented on src/nvme/mi.h:
```diff
@@ -1020,6 +1190,89 @@ static inline int nvme_mi_mi_config_set_mctp_mtu(nvme_mi_ep_t ep, __u8 port,
+				struct nvme_mi_ae_occ_list_hdr *occ_list,
+				size_t *occ_list_num_bytes);
+
+static inline int aem_ack(nvme_mi_ep_t ep,
```
prefixed name here

chorkin replied:
It's a static function.  What is the point of the prefix?

jk-ozlabs commented on src/nvme/mi.h:
```diff
@@ -3086,4 +3339,157 @@ int nvme_mi_admin_format_nvm(nvme_mi_ctrl_t ctrl,
+ * nvme_mi_aem_get_enabled() - Return information on which AEs are enabled
+ * @ep: Endpoint to check enabled status
+ * @enabled: buffer indexed by AE event ID of enabled state
+ * @len: length of enabled.  Must be 256 bytes.
```
The call semantics here are a bit odd; you're referring to `bytes` in the comment, but it's 256 booleans (which probably corresponds to 256 bytes, but that's not guaranteed).

We could define a more compact type for this perhaps; something like

```c
typedef uint64_t nvme_mi_aem_mask_t[4]
```

or if you prefer the boolean approach:

```c
typedef bool nvme_mi_aem_mask_t[256]
```

then drop the len argument, because you have a specific type instead.

Open to other suggestions though :)

jk-ozlabs commented on examples/mi-mctp-ae.c:
```diff
@@ -0,0 +1,188 @@
+#include <ccan/endian/endian.h>
+#include <sys/select.h>
+
+// Function to print the byte array
```
I reckon we can deduce that from the function name :)

jk-ozlabs commented on examples/mi-mctp-ae.c:
```diff
@@ -0,0 +1,188 @@
+	if (event->spec_info_len && event->spec_info)
+		print_byte_array(event->spec_info, event->spec_info_len);
+	else
+		printf("EMPTY\n");
```
Maybe just skip the `specific_info` message instead of reporting the thing as empty.

jk-ozlabs commented on examples/mi-mctp-ae.c:
```diff
@@ -0,0 +1,188 @@
+
+	printf("Press any key to exit\n");
+	while (1) {
+		int poll_timeout = 500; // Timeout in milliseconds
```
Now that you have unified the polling, do you need a timeout at all?

jk-ozlabs commented on src/nvme/mi-mctp.c:
```diff
@@ -65,6 +66,7 @@ struct sockaddr_mctp {
 	__u8			__smctp_pad1;
 };
 
+
```
Not needed

jk-ozlabs commented on src/nvme/mi-mctp.c:
```diff
@@ -84,6 +86,9 @@ struct nvme_mi_transport_mctp {
 	int	sd;
 	void	*resp_buf;
 	size_t	resp_buf_size;
+	int		sd_aem;
```
alignment seems odd here

jk-ozlabs replied:
Still seems odd; `sd_aem` looks to be one indent too much?

(super minor, but still)

jk-ozlabs replied:
Still incorrect alignment. You have:

```c
	void	*resp_buf;
	size_t	resp_buf_size;
	int		sd_aem;
```

you want:

```c
	void	*resp_buf;
	size_t	resp_buf_size;
	int	sd_aem;
```

There are other instances of the same thing elsewhere too. I won't worry about this for review, but it's still a minor niggle.

chorkin commented on src/nvme/mi.c:
```diff
@@ -1974,3 +2131,434 @@ const char *nvme_mi_status_to_string(int status)
+	__u8 aemd,
+	__u8 aerd,
+	struct nvme_mi_aem_callbacks *callbacks,
+	void *userdata)
```
Yes.  We can do that.  It's stored in Read NVMe-MI Data Structure.  Although, I suspect a CONFIGURATION GET for AEM type should return an error if it's not supported.  So this should return an error anyways before attempting to enable.  Do you think it's worth the overhead of adding an extra NVME-MI call in this API so that we can characterize the error a bit more precisely?

jk-ozlabs replied:
OK, I think just handling the detection through the setup path would be fine then. As long as we're handling the error properly there.

chorkin commented on src/nvme/mi.c:
```diff
@@ -1974,3 +2131,434 @@ const char *nvme_mi_status_to_string(int status)
+
+	//It's possible we're already enabled
+	if (!ep->aem_ctx)
+		ep->aem_ctx = malloc(sizeof(*ep->aem_ctx));
```
The idea was to leave a clue about if we currently have the AEMs enabled or not.  It wasn't about being inconsistent.

chorkin commented on src/nvme/mi.h:
```diff
@@ -1020,6 +1190,89 @@ static inline int nvme_mi_mi_config_set_mctp_mtu(nvme_mi_ep_t ep, __u8 port,
+	list.hdr.aeetl = sizeof(struct ae_enable_list_header_t);
+	list.hdr.numaee = 0;
+
+	return nvme_mi_mi_config_set_async_event(ep,
+										false,
+										false,
+										false,
+										0,
+										0,
+										&list,
+										sizeof(list),
+										occ_list,
+										occ_list_num_bytes);
```
I'm not sure what.  I have these each on new lines and indented to be matching indentation....Trying to get past checkpatch issues for long lines

jk-ozlabs replied:
Seems related to the tab width above. This could either be indented as you have done (but correctly aligned):

```c

	return nvme_mi_mi_config_set_async_event(ep,
						 false,
						 false,
						 false,
						 0,
						 0,
						 &list,
						 sizeof(list),
						 occ_list,
						 occ_list_num_bytes);
```

however, since those parameters are all pretty small, it's probably better to compact this to 80/100 cols:

```c
	return nvme_mi_mi_config_set_async_event(ep, false, false, false, 0, 0,
						 &list, sizeof(list), occ_list,
						 occ_list_num_bytes);
```

chorkin commented:
@jk-ozlabs, I think I've addressed most of the comments, although still looking for a bit of guidance on what to do instead of using the bitmapped structs.

I also pushed tests for the features which I believe are pretty extensive and helped uncover a few bugs that existed.

Cheers!

jk-ozlabs commented:
> although still looking for a bit of guidance on what to do instead of using the bitmapped structs.

sure!

So you currently have this:

```c
/**
 * struct nvme_mi_aem_supported_item - AE Supported List Item
 * @aesl: AE supported list item length
 * @aesi: AE supported info
 *
 * Following this header should be hdr.numaes entries of
 * nvme_mi_aem_supported_item structures
 */
struct nvme_mi_aem_supported_item {
	__u8 aesl;//Length of this item.  Set to 3
	struct {
		__le16 aeis     : 8; //Identifier of supported ae
		__le16 reserved : 7;
		__le16 aese     : 1; //AE Support Enabled
	} __attribute__((packed)) aesi;
} __attribute__((packed));
_Static_assert(sizeof(struct nvme_mi_aem_supported_item) == 3,
	"sizeof_nvme_mi_aem_supported_item_is_not_3_bytes");
```

I would suggest dropping the sub-byte fields here, so that you only have whole-byte members, which the standard still provides defined names for.

This would give you something like:

```c
struct nvme_mi_aem_supported_item {
	__u8 aesl;
	__le16 aesi;
} __attribute__((packed));
```

You may want to provide some convenience accessors for the sub-fields:

```c
static inline bool nvme_mi_aem_aesi_get_aese(__le16 aesi)
{
	return !!(le16_to_cpu(aesi) & 0x8000);
}

static inline u8 nvme_mi_aem_aesi_get_aesi(__le16 aesi)
{
	return le16_to_cpu(aesi) & 0xff;
}
```

(where I'm proposing `bool` for single-bit types, but you may have other preferences. We could equivalently just use an `unsigned int` for all values if that makes more sense)

The AEMTI field requires shifting too, so something like:

```c
struct nvme_mi_aem_occ_list_hdr {
	__u8 numaeo;
	__u8 aelver;
	__let16 aeolli;
	__u8 aeolhl;
	__u8 aemti;
} __attribute__((packed));

static inline bool nvme_mi_aem_aemti_get_aemgn(__u8 aemti)
{
	return aemti >> 3 & 0x1f;
}
```

You also have the `NVME_GET` and `NVME_SET` macros available, if that helps with defining these. Or, perhaps even better, not defining  the individual accessors at all, and just using GET/SET at the individual callsites. You'll need to define `_MASK` and `_SHIFT` for each field for those though.

Also, I would suggest moving the `_Static_asserts` out of the header and into mi.c; there may be compiler compatibility issues, which we would avoid imposing on the library user here (`_Static_assert` is deprecated in C23, but `static_assert` is only available in C23...)

jk-ozlabs commented on src/nvme/private.h:
```diff
@@ -289,9 +305,11 @@ __u32 nvme_mi_crc32_update(__u32 crc, void *data, size_t len);
+	int (*aem_socket)(__u8 eid, unsigned int network);
 	ssize_t (*sendmsg)(int, const struct msghdr *, int);
 	ssize_t (*recvmsg)(int, struct msghdr *, int);
+	ssize_t (*recv)(int __fd, void *__buf, size_t __n, int __flags);
```
Are you sure we need this? `recvmsg` behaviour is a superset of `recv`

(granted, you do need to define a `struct msghdr` in the caller, but that should be less intrusive than a whole new callback)

jk-ozlabs commented on src/nvme/mi-mctp.c:
```diff
@@ -220,6 +254,160 @@ static bool nvme_mi_mctp_resp_is_mpr(void *buf, size_t len,
+	struct nvme_mi_transport_mctp *mctp = ep->transport_data;
+
+	// Read until there is no more data
+	ops.recv(mctp->sd_aem, &buffer, sizeof(buffer), MSG_TRUNC);
```
[I have commented elsewhere on recv vs. recvmsg]

I think you'll still need the loop here; since this is a datagram socket you will only get one message per call.

Just for my understanding though: we're only calling this after disabling AEMs, right? In which case there's no race between purged and new messages.

Also, you may be able to get away with no `buffer` at all, and just passing a zero-item iovec. I haven't tried that, but I assume that that will still work with `MSG_TRUNC`.

chorkin replied:
Yes.  That's correct.  After the disable.

chorkin commented:
> Where are we this PR (btw, still in `ready to review` state)?

Thanks for the callout @igaw .  I forgot we were in draft state.  I have no pending changes from my side.  Waiting for any remaining feedback, PLUS we still need the new kernel bits.

jk-ozlabs commented on src/nvme/mi.c:
```diff
@@ -414,6 +432,66 @@ static int nvme_mi_verify_resp_mic(struct nvme_mi_resp *resp)
+	if (nvme_mi_ep_has_quirk(ep, NVME_QUIRK_MIN_INTER_COMMAND_TIME))
+		nvme_mi_record_resp_time(ep);
+
+	if (rc && errno == EWOULDBLOCK) {
```
`errno` may be overwritten by the `clock_gettime` call in `nvme_mi_record_resp_time` there. You may want to save it, or shift the record_resp_time call.

Given the read is non-blocking, you should be fine to record the time on function entry.

jk-ozlabs commented on src/nvme/mi-mctp.c:
```diff
@@ -84,15 +85,47 @@ struct nvme_mi_transport_mctp {
+	addr.smctp_family = AF_MCTP;
+	addr.smctp_network = network;
+	addr.smctp_addr.s_addr = MCTP_ADDR_ANY;
+	addr.smctp_type = MCTP_TYPE_NVME | MCTP_TYPE_MIC;
```
The MIC bit is ignored in the bind, just make this `MCTP_TYPE_NVME`

jk-ozlabs commented on src/nvme/mi-mctp.c:
```diff
@@ -84,15 +85,47 @@ struct nvme_mi_transport_mctp {
+	if (sd < 0)
+		return sd;
+
+	memset(&addr, 0, sizeof(addr));
```
Super minor: you have already set to `{0}` above.

jk-ozlabs commented on src/nvme/mi-mctp.c:
```diff
@@ -487,13 +690,26 @@ nvme_mi_ep_t nvme_mi_open_mctp(nvme_root_t root, unsigned int netid, __u8 eid)
+		goto err_free_aem_rspbuf;
+	}
+
+	mctp->sd_aem = ops.aem_socket(eid, netid);
+	if (mctp->sd_aem < 0) {
+		errno_save = errno;
+		goto err_close_sd;
```
The aem socket creation may fail if the kernel does not have support for the per-EID AEM sockets (which will be likely when this first hits upstream). We don't want that to prevent all MI endpoint creation though, so a failure here should not fail the open.

A couple of options:
 - only attempt to open the sd_aem when AEMs are requested through `nvme_mi_aem_enable` (ie, where you're creating the `aem_ctx` too)
 - try here, but all other uses of `sd_aem` check whether the socket is valid first

I think the former may be cleaner

chorkin replied:
Moved initialization into nvme_mi_aem_enable per suggestion

jk-ozlabs reviewed (commented):
Looking good, a few comments, mostly from the reworks, and how we handle errors on `sd_aem` creation.

jk-ozlabs commented:
For handling the upcoming kernel support, I'm thinking something along these lines:

```diff
diff --git a/src/nvme/mi-mctp.c b/src/nvme/mi-mctp.c
index 859503d5..c21cb8a1 100644
--- a/src/nvme/mi-mctp.c
+++ b/src/nvme/mi-mctp.c
@@ -102,25 +102,36 @@ static int nvme_mi_msg_socket(void)
 
 static int nvme_mi_aem_socket(__u8 eid, unsigned int network)
 {
-       struct sockaddr_mctp addr =  {0};
-       int sd = socket(AF_MCTP, SOCK_DGRAM|SOCK_NONBLOCK, 0);
+       struct sockaddr_mctp local_addr = {0}, remote_addr = {0};
+       int sd, rc;
 
+       sd = socket(AF_MCTP, SOCK_DGRAM | SOCK_NONBLOCK, 0);
        if (sd < 0)
                return sd;
 
-       memset(&addr, 0, sizeof(addr));
-       addr.smctp_family = AF_MCTP;
-       addr.smctp_network = network;
-       addr.smctp_addr.s_addr = MCTP_ADDR_ANY;
-       addr.smctp_type = MCTP_TYPE_NVME | MCTP_TYPE_MIC;
-       addr.smctp_tag = MCTP_TAG_OWNER;
+       remote_addr.smctp_family = AF_MCTP;
+       remote_addr.smctp_network = network;
+       remote_addr.smctp_addr.s_addr = eid;
+       remote_addr.smctp_type = MCTP_TYPE_NVME;
+       /* connect() will specify a remote EID for the upcoming bind() */
+       rc = connect(sd, (struct sockaddr *)&remote_addr, sizeof(remote_addr));
+       if (rc)
+               goto err_close;
 
-       if (bind(sd, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
-               close(sd);
-               return -1;
-       }
+       local_addr.smctp_family = AF_MCTP;
+       local_addr.smctp_network = network;
+       local_addr.smctp_addr.s_addr = MCTP_ADDR_ANY;
+       local_addr.smctp_type = MCTP_TYPE_NVME;
+
+       rc = bind(sd, (struct sockaddr *)&local_addr, sizeof(local_addr));
+       if (rc)
+               goto err_close;
 
        return sd;
+
+err_close:
+       close(sd);
+       return -1;
 }
```

igaw commented on src/nvme/mi.h:
```diff
@@ -86,6 +86,7 @@
 
 #include <endian.h>
 #include <stdint.h>
+#include <ccan/endian/endian.h>
```
Yep, we should not expose these internal headers.

chorkin replied:
Done.

chorkin commented on src/nvme/mi.h:
```diff
@@ -307,6 +311,217 @@ enum nvme_mi_config_smbus_freq {
+ * This header preceeds a number, (&numaes), of AE supported data structures
+ */
+struct nvme_mi_aem_supported_list_header {
+	__u8 numaes; //Should be zero
```
Good catch.  This was on the wrong row

chorkin commented:
> For handling the upcoming kernel support, I'm thinking something along these lines:
> 
> ```diff
> diff --git a/src/nvme/mi-mctp.c b/src/nvme/mi-mctp.c
> index 859503d5..c21cb8a1 100644
> --- a/src/nvme/mi-mctp.c
> +++ b/src/nvme/mi-mctp.c
> @@ -102,25 +102,36 @@ static int nvme_mi_msg_socket(void)
>  
>  static int nvme_mi_aem_socket(__u8 eid, unsigned int network)
>  {
> -       struct sockaddr_mctp addr =  {0};
> -       int sd = socket(AF_MCTP, SOCK_DGRAM|SOCK_NONBLOCK, 0);
> +       struct sockaddr_mctp local_addr = {0}, remote_addr = {0};
> +       int sd, rc;
>  
> +       sd = socket(AF_MCTP, SOCK_DGRAM | SOCK_NONBLOCK, 0);
>         if (sd < 0)
>                 return sd;
>  
> -       memset(&addr, 0, sizeof(addr));
> -       addr.smctp_family = AF_MCTP;
> -       addr.smctp_network = network;
> -       addr.smctp_addr.s_addr = MCTP_ADDR_ANY;
> -       addr.smctp_type = MCTP_TYPE_NVME | MCTP_TYPE_MIC;
> -       addr.smctp_tag = MCTP_TAG_OWNER;
> +       remote_addr.smctp_family = AF_MCTP;
> +       remote_addr.smctp_network = network;
> +       remote_addr.smctp_addr.s_addr = eid;
> +       remote_addr.smctp_type = MCTP_TYPE_NVME;
> +       /* connect() will specify a remote EID for the upcoming bind() */
> +       rc = connect(sd, (struct sockaddr *)&remote_addr, sizeof(remote_addr));
> +       if (rc)
> +               goto err_close;
>  
> -       if (bind(sd, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
> -               close(sd);
> -               return -1;
> -       }
> +       local_addr.smctp_family = AF_MCTP;
> +       local_addr.smctp_network = network;
> +       local_addr.smctp_addr.s_addr = MCTP_ADDR_ANY;
> +       local_addr.smctp_type = MCTP_TYPE_NVME;
> +
> +       rc = bind(sd, (struct sockaddr *)&local_addr, sizeof(local_addr));
> +       if (rc)
> +               goto err_close;
>  
>         return sd;
> +
> +err_close:
> +       close(sd);
> +       return -1;
>  }
> ```

Just curious.  Why do we need the network in both the locall_addr and the remote_addr?  Are they actually both used?

jk-ozlabs commented:
I had a reply typed-out for that, but looks like it didn't send! Here's a re-do:

> Just curious. Why do we need the network in both the locall_addr and the remote_addr? Are they actually both used?

Great question. We need the networks to match across the connect and bind calls, but we can specify `MCTP_NET_ANY` for the bind.

The `connect()` call specifies exactly one remote address, by EID. Since EIDs are net-specific, we need to specify a net there. Like the existing `sendto`, a value of `MCTP_NET_ANY` for `connect()` there is a default, and would get resolved to the system-wide "default" MCTP network - rather than specifying any net. This makes single-network system configurations easier to handle.

The `bind()` call then flags the socket for receiving incoming (TO=1) messages, and we're specifying a local address here. We want the nets to correspond, because otherwise it wouldn't make sense to be bound to a net that the remote endpoint (whose address was specified in the `connect()`) is not present on.

We can also use `MCTP_NET_ANY` here, for the bind() to automatically match the `connect`, but we have the value used anyway, so we just re-use that.

jk-ozlabs commented:
We have an initial cut of the kernel support for the peer-specific binds, in a `dev/mctp-peer-bind` branch:

https://github.com/CodeConstruct/linux/commits/dev/mctp-peer-bind/

Are you able to give this a test run on your side?

chorkin commented:
> We have an initial cut of the kernel support for the peer-specific binds, in a `dev/mctp-peer-bind` branch:
> 
> https://github.com/CodeConstruct/linux/commits/dev/mctp-peer-bind/
> 
> Are you able to give this a test run on your side?

Hi @jk-ozlabs , as I mentioned previously, we're using a QEMU instance built by a partner special to do this work.  I'm going to reach out and see if they're willing to generate a build for us, but not sure if they will.  Even if they do, I'll only be able to test this with a single endpoint based on how our QEMU system works, so I won't be able to test the new parts that allow multiple bindings.  Nonetheless, let me reach out and see what I can accomplish here to unblock this.

chorkin commented:
> > We have an initial cut of the kernel support for the peer-specific binds, in a `dev/mctp-peer-bind` branch:
> > https://github.com/CodeConstruct/linux/commits/dev/mctp-peer-bind/
> > Are you able to give this a test run on your side?
> 
> Hi @jk-ozlabs , as I mentioned previously, we're using a QEMU instance built by a partner special to do this work. I'm going to reach out and see if they're willing to generate a build for us, but not sure if they will. Even if they do, I'll only be able to test this with a single endpoint based on how our QEMU system works, so I won't be able to test the new parts that allow multiple bindings. Nonetheless, let me reach out and see what I can accomplish here to unblock this.

@jk-ozlabs , I wanted to follow up here.  Sounds like we're going to be enabled to build kernel changes so I will test this.  Do I just need to copy your example code as is to do it or are there other things I'll need to do?  How do the kernel bindings work?  Are they runtime or some kind of header defines based?

Thanks.

jk-ozlabs commented:
> Sounds like we're going to be enabled to build kernel changes so I will test this

super!

> Do I just need to copy your example code as is to do it or are there other things I'll need to do?

You'll just need to:

 1. incorporate that `mctp-peer-bind` branch in to your kernel tree.
 2. Add the diff I included above to libnvme

for (1), easiest approach (if you already have a custom kernel) would be to grab commit https://github.com/CodeConstruct/linux/commit/2e9c0d52758f9c4dbbd26a5993d05276c3e58d32 and onwards, and apply those to your custom kernel. Or, just build that `mctp-peer-bind` branch directly if you have no other changes that you need

for (2) feel free to incorporate that into your change, or I can provide as an actual commit if you prefer.

chorkin commented on src/libnvme-mi.map:
```diff
@@ -1,3 +1,13 @@
+LIBNVME_MI_1_13 {
+	global:
+		nvme_mi_aem_get_fd;
```
We missed 1.13.  Need to bump this up to 1_14

jk-ozlabs replied:
@igaw any thoughts on using an `UNRELEASED` group for these?

If so, we'd just assign the actual version number as part of the release commit. Also means that @chorkin doesn't need to predict upcoming release numbers (say, if there was a new major required from a future change in the same group)

igaw replied:
Yes, 1.13 is already out. We could start doing this and update the `scripts/release.sh` script to update the section before doing the release. If it is automated I am fine with it. I just know I will forget it to do it eventually, so it needs to be scripted :)

igaw replied:
has anyone experience with awk? https://github.com/linux-nvme/libnvme/pull/1001

jk-ozlabs replied:
> has anyone experience with awk? #1001

yes, unfortunately. I'll comment there.

chorkin commented:
> > Sounds like we're going to be enabled to build kernel changes so I will test this
> 
> super!
> 
> > Do I just need to copy your example code as is to do it or are there other things I'll need to do?
> 
> You'll just need to:
> 
> 1. incorporate that `mctp-peer-bind` branch in to your kernel tree.
> 2. Add the diff I included above to libnvme
> 
> for (1), easiest approach (if you already have a custom kernel) would be to grab commit [CodeConstruct/linux@2e9c0d5](https://github.com/CodeConstruct/linux/commit/2e9c0d52758f9c4dbbd26a5993d05276c3e58d32) and onwards, and apply those to your custom kernel. Or, just build that `mctp-peer-bind` branch directly if you have no other changes that you need
> 
> for (2) feel free to incorporate that into your change, or I can provide as an actual commit if you prefer.

Alrighty!  I was able to build our QEMU today and test with your commits on top.  I did two tests:

#1) Without your changes, ran with the new peer-bind code in libnvme-mi, and verified an error is returned in the example app since this is not supported yet
#2) With your changes, did the same test.  Verified no error and verified I can trigger AEs which show up in my example app.

What I'm not able to test is multiple endpoints with AEs because that's just not something this QEMU setup will allow at this point, but if you are confident in your kernel changes, or can test that bit on your end, I think we're good to go!

Cheers and have a great weekend!

jk-ozlabs commented:
nice one!

The multiple-endpoints functionality (at the MCTP level) is something we have tests for, so I'm fine to cover that on our side.

[We have a facility for multiple NVMe-MI-device emulation under qemu, but the device-side doesn't support AEs at present.  I'll give this a go once that is implemented though]

chorkin commented:
@jk-ozlabs are we good to complete this, then?

jk-ozlabs commented on examples/mi-mctp-ae.c:
```diff
@@ -0,0 +1,183 @@
+
+	ep = nvme_mi_open_mctp(root, net, eid);
+	if (!ep)
+		errx(EXIT_FAILURE, "can't open MCTP endpoint %d:%d", net, eid);
```
`errno` should be set here, you may want to use `err` instead of `errx`.

jk-ozlabs commented on examples/mi-mctp-ae.c:
```diff
@@ -0,0 +1,183 @@
+
+	rc = nvme_mi_aem_get_enabled(ep, &enabled_map);
+	if (rc)
+		errx(EXIT_FAILURE, "Can't query enabled aems:%d (%d)", rc, errno);
```
Given want errno, just use `err` here and drop the explicit `errno` format.

(as well as the other calls where you format errno below too)

jk-ozlabs commented on examples/mi-mctp-ae.c:
```diff
@@ -0,0 +1,183 @@
+
+	rc = nvme_mi_aem_enable(ep, &aem_config, &data);
+	if (rc && errno == EOPNOTSUPP)
+		errx(EXIT_FAILURE, "MCTP Peer-Bind is required for AEM\n");
```
no need for `\n` with `err` / `warn`

(and below also)

jk-ozlabs commented on examples/mi-mctp-ae.c:
```diff
@@ -0,0 +1,183 @@
+		rc = poll(fds, 2, -1);
+
+		if (rc == -1) {
+			perror("poll");
```
maybe `warn("poll")` would be more consistent here

jk-ozlabs commented on src/nvme/mi-mctp.c:
```diff
@@ -459,8 +671,34 @@ static const struct nvme_mi_transport nvme_mi_transport_mctp = {
+	mctp = ep->transport_data;
+
+	//This doesn't have to be done multiple times
+	if (mctp->sd_aem > 0)
```
`0` is a valid fd (although unlikely), you probably want `>= 0` here.

jk-ozlabs commented on src/nvme/mi-mctp.c:
```diff
@@ -433,7 +640,12 @@ static void nvme_mi_mctp_close(struct nvme_mi_ep *ep)
+	if (mctp->sd_aem > 0)
+		close(mctp->sd_aem);
+	if (ep->aem_ctx)
+		free(ep->aem_ctx);
```
`free(NULL)` is well-defined; you don't really need the conditional here.

jk-ozlabs commented on src/nvme/mi-mctp.c:
```diff
@@ -220,6 +264,169 @@ static bool nvme_mi_mctp_resp_is_mpr(void *buf, size_t len,
+		goto out;
+	}
+
+	/* Start unpacking the linear resp buffer into the split header + data
+	 * + MIC. We check for a MPR response before fully unpacking, as we'll
+	 * need to preserve the resp layout if we need to retry the receive.
```
Is the second sentence still valid? We won't be getting an MPR response here.

jk-ozlabs commented on src/nvme/mi-mctp.c:
```diff
@@ -433,7 +640,12 @@ static void nvme_mi_mctp_close(struct nvme_mi_ep *ep)
 
 	mctp = ep->transport_data;
 	close(mctp->sd);
+	if (mctp->sd_aem > 0)
```
`>= 0`, but you could probably just `close()` unconditionally if you prefer.

jk-ozlabs commented on test/mi-mctp.c:
```diff
@@ -713,6 +741,655 @@ static void test_mpr_mprt_zero(nvme_mi_ep_t ep, struct test_peer *peer)
 	assert(rc == 0);
 }
 
+enum AemEnableState {
+	AemEnableGetEnabled,
+	AemSetToDisabled,
+	AemEnableSetEnabled,
+	AemProcess,
+	AemAckResponse,
+	AemAckReceived
+};
+
+enum AemFailureCondition {
+	AemFcNone,
+	AemFcBadGetConfigHeaderLen,
+	AemFcBadGetConfigTotalLen,
+	AemFcBadGetConfigBufferLen,
+	AemFcBadOccRspHdrLen_Sync,
+	AemFcBadOccRspTotalLen_Sync,
+	AemFcBadOccRspBufferLen_Sync,
+	AemFcBadOccRspHdrLen_Aem,
+	AemFcBadOccRspTotalLen_Aem,
+	AemFcBadOccRspBufferLen_Aem,
+};
```
These are a little inconsistent with existing coding style; if this is to match up with some other definitions we can probably let it slide, being test code. If not though, can you make these consistent with the rest here? mainly just `snake_case`.

jk-ozlabs commented on test/mi.c:
```diff
@@ -95,6 +95,10 @@ static const struct nvme_mi_transport test_transport = {
 	.submit = test_transport_submit,
 	.close = test_transport_close,
 	.desc_ep = test_transport_desc_ep,
+	//TO DO:
```
what is yet to do here? expanding on this will help when we find this in three years and forget what we're meant to do :)

chorkin replied:
I think this is stale.  I'm not sure there is more work here, and really these are just not needed to be implemented.  I updated the comment.

jk-ozlabs commented on test/mi-mctp.c:
```diff
@@ -713,6 +741,655 @@ static void test_mpr_mprt_zero(nvme_mi_ep_t ep, struct test_peer *peer)
+
+enum nvme_mi_aem_handler_next_action aem_handler(nvme_mi_ep_t ep, size_t num_events, void *userdata)
+{
+	struct aem_rcv_enable_fn_data *fn_data = (struct aem_rcv_enable_fn_data *)userdata;
```
Note that you don't need to cast to/from `void *`.

This means we can avoid a lot of casts, where casts are something we want to be careful about, and typically need a more thorough review.

chorkin replied:
Did not realize that.  This was habitual.

jk-ozlabs commented on src/nvme/mi.h:
```diff
@@ -307,6 +310,229 @@ enum nvme_mi_config_smbus_freq {
+ * @aesid: aesid value to use
+ * @item: Pointer to @nvme_mi_aem_supported_item to update the aesi field
+ */
+void nvme_mi_aem_aesi_set_aesid(__u8 aesid, struct nvme_mi_aem_supported_item *item);
```
I'd tend to have the modified thing (the "context") as the first argument.

(Also for the `_set` functions below too)

jk-ozlabs commented on src/nvme/mi.h:
```diff
@@ -1020,6 +1254,81 @@ static inline int nvme_mi_mi_config_set_mctp_mtu(nvme_mi_ep_t ep, __u8 port,
+ * @enable_list: nvme_mi_aem_enable_listucture containing header and items
+ * of events to be enabled or disabled.  This is taken as a delta change
+ * from the current configuration.
+ * @enable_list_length: Length of the enable_list in bytes including header and data.
+ * Meant to catch overrun issues.
+ * @occ_list: Pointer to populate with the occurrence list (header and data)
+ * @occ_list_num_bytes: Total length of provided occ_list buffer in bytes.  Will be
+ * updated with received size if successful
```
Minor inconsistency in naming here: the `enable_list_length` vs. `occ_list_num_bytes`.

A `_length` typically represents a number of elements, but it's a size (you mention "in bytes" in the comment). `_num_bytes` properly describes the latter, but `_size` is more conventional if properly commented.

I'd suggest `_size` for both, if we're working with sizes in bytes.

jk-ozlabs reviewed (commented):
Looking good!

I have done a bit of a nit-picky review here; most is not super important.

However: we will want to make sure we have the `mi.h` changes correct, as they'll become very difficult to modify once merged. If nothing else take a look through those and let me know if you disagree on any points there.

Thanks for all the work on this!

jk-ozlabs reviewed (approved):
The API all looks good to me!

chorkin commented:
@igaw , this is ready to merge.

igaw commented:
Oh, it's a non trivial rebase. @chorkin could you rebase it against the current master, I think I'll break a thing or two if I do it myself. Sorry.

chorkin commented:
I need to do one final check, @igaw .  Will post here when done.

chorkin commented:
> I need to do one final check, @igaw . Will post here when done.

@igaw, rebased and ready to merge.

igaw commented:
Thanks!

jk-ozlabs commented:
🥳

chorkin commented:
Thanks everyone!  This is such a cool new feature; I'm glad we could get it in!
